src/apprt/gtk/close_dialog.zig

```diff
@@ -0,0 +1,173 @@
+        parent: Parent.Class,
+
+        pub const Instance = CloseDialog;
+    };
+
+    /// The target of this close dialog.
+    ///
+    /// This is here so that we can consolidate all logic related to
+    /// prompting the user and closing windows/tabs/surfaces/etc.
+    /// together into one struct that is the sole source of truth.
+    pub const Target = union(enum) {
+        app: *App,
+        window: *Window,
+        tab: *Tab,
+        surface: *Surface,
+
+        pub fn title(self: Target) [:0]const u8 {
+            return switch (self) {
+                .app => "Quit Ghostty?",
+                .window => "Close Window?",
+                .tab => "Close Tab?",
+                .surface => "Close Surface?",
+            };
+        }
+
+        pub fn body(self: Target) [:0]const u8 {
+            return switch (self) {
+                .app => "All terminal sessions will be terminated.",
+                .window => "All terminal sessions in this window will be terminated.",
+                .tab => "All terminal sessions in this tab will be terminated.",
+                .surface => "The currently running process in this surface will be terminated.",
+            };
+        }
+
+        pub fn dialogWindow(self: Target) ?*gtk.Window {
+            return switch (self) {
+                .app => {
+                    // Find the currently focused window. We don't store this
+                    // anywhere inside the App structure for some reason, so
+                    // we have to query every single open window and see which
+                    // one is active (focused and receiving keyboard input)
+                    const list = gtk.Window.listToplevels();
+                    defer list.free();
+
+                    const focused = list.findCustom(null, findActiveWindow);
+                    return @ptrCast(@alignCast(focused.f_data));
+                },
+                .window => |v| @ptrCast(v.window),
+                .tab => |v| @ptrCast(v.window.window),
+                .surface => |v| surface: {
+                    const window_ = v.container.window() orelse return null;
+                    break :surface @ptrCast(window_.window);
+                },
+            };
+        }
+
+        fn close(self: Target) void {
+            return switch (self) {
+                .app => |v| v.quitNow(),
+                .window => |v| gtk.Window.destroy(@ptrCast(v.window)),
+                .tab => |v| v.remove(),
+                .surface => |v| v.container.remove(),
+            };
+        }
+    };
+};
+
+fn findActiveWindow(data: ?*const anyopaque, _: ?*const anyopaque) callconv(.C) c_int {
+    const window: *gtk.Window = @ptrCast(@alignCast(@constCast(data orelse return -1)));
+
+    // Confusingly, `isActive` returns 1 when active,
+    // but we want to return 0 to indicate equality.
+    // Abusing integers to be enums and booleans is a terrible idea, C.
+    return if (window.isActive() > 0) 0 else -1;
```

Comment:
`window.isActive()` returns an int in zig-gobject? I also don't love the `> 0`. Why not just compare to `1` or `c.TRUE`.

Reply:
I mean, you could write it as `@intFromBool(!(window.isActive() == 0))` but that's just even harder to understand what it does

Reply:
> I mean, you could write it as `@intFromBool(!(window.isActive() == 0))` but that's just even harder to understand what it does. GLib's boolean semantics are "0 if false, anything else if true", so it's theoretically possible that a truthy value won't be exactly 1.

If you want to be pedantic, sure, butthe function is documented as returning `TRUE`: https://docs.gtk.org/gtk4/method.Window.is_active.html

Reply:
I generally prefer `!= 0` when comparing C truthy values, as `0` is `false` and anything else is `true`.